### ISSUE TITLE
auto-direct-node-routes with clustermesh makes cilium crash after updating from 1.14 to 1.15

### ISSUE THREAD

**Issue by lgyurci** [cilium/cilium#31843]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

After updating from 1.14 to 1.15, the cilium agent crashes.

The cause of the crash is, that cilium is trying to install remote nodes (the members of other clusters) as gateways into the hosts routing table (since auto-direct-node-routes is set to true), which does not succeed, since they are not on the same LAN. Cilium 1.14 also tried to install these nodes as gateways for their respective pod ranges, but it silently failed, and only spit out a warning (as seen below). We use native routing, with dsr. The network fabric between the clusters is able to route the pod IPs between the clusters, and our setup has worked in 1.14.4.

### Cilium Version

Updated from 1.14.4 to 1.15.1

### Kernel Version

5.14.0

### Kubernetes Version

v1.26.15+rke2r1

### Regression

1.14

### Relevant log output

```shell
--- IN 1.14.4 ---
cilium-agent level=warning msg="Unable to install direct node route {Ifindex: 0 Dst: 10.145.1.0/24 Src: <nil> Gw: 10.176.50.6 Flags: [] Table: 0 Realm: 0}" error="route to destination 10.176.50.6 contains gateway 10.156.50.254, must be directly reachable" subsys=linux-datapath                                                                                                                                                                                                   
cilium-agent level=warning msg="Unable to install direct node route {Ifindex: 0 Dst: 10.145.2.0/24 Src: <nil> Gw: 10.186.50.25 Flags: [] Table: 0 Realm: 0}" error="route to destination 10.186.50.25 contains gateway 10.156.50.254, must be directly reachable" subsys=linux-datapath                                                                                                                                                                                                 
cilium-agent level=warning msg="Unable to install direct node route {Ifindex: 0 Dst: 10.175.1.0/24 Src: <nil> Gw: 10.176.50.37 Flags: [] Table: 0 Realm: 0}" error="route to destination 10.176.50.37 contains gateway 10.156.50.254, must be directly reachable" subsys=linux-datapath                                                                                                                                                                                                 
cilium-agent level=warning msg="Unable to install direct node route {Ifindex: 0 Dst: 10.175.0.0/24 Src: <nil> Gw: 10.176.50.38 Flags: [] Table: 0 Realm: 0}" error="route to destination 10.176.50.38 contains gateway 10.156.50.254, must be directly reachable" subsys=linux-datapath                                                                                                                                                                                                 
cilium-agent level=warning msg="Unable to install direct node route {Ifindex: 0 Dst: 10.175.2.0/24 Src: <nil> Gw: 10.176.50.36 Flags: [] Table: 0 Realm: 0}" error="route to destination 10.176.50.36 contains gateway 10.156.50.254, must be directly reachable" subsys=linux-datapath                                                                                                                                                                                                 

--- IN 1.15.1 ---
cilium-agent level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: failed to enable direct routes for ipv4: route to destination 10.156.50.38 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.37 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.36 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.186.50.25 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.5 contains gateway 10.176.50.254, must be directly reachable" subsys=daemon
```

### Anything else?

The above 2 logs are recorded on two different clusters

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by lmb** [comment]
Can you provide us with a sysdump / helm values etc?

**Comment by giorio94** [comment]
AFAIK, this scenario is not officially supported at the moment, and `auto-direct-node-routes` applies to both nodes in the local, and possible remote clusters. In other words, there isn't currently the possibility of applying that to a subset of nodes only (although I see that could be useful in certain scenarios). That was already the case in v1.14, although that didn't only triggered a warning message, and didn't cause a hard failure. 

That said, and independently of this specific use-case, the behavioral change appears to have introduced in d890e629f2fd891986166671a47108cf74f7f607 as an effort of bubbling up more errors, although it is not totally clear to me whether the extreme consequence of panic'ing the agent was intended or not. @tommyp1ckles  could you please chime in with your thoughts?

**Comment by lgyurci** [comment]
The thing that is strange to me in this question, is when would you actually want auto-direct-node-routes to import remote-nodes as gateways? Only when you have the entire clustermesh on a layer 2 network, which is very rarely the case. The "fail silently" was great in our case (even better than excluding remote nodes entirely), because we have some nodes from the other cluster on the same LAN, so those can communicate without going packets going through the entire network stack. Also, after this fatal error, cilium throws in a segmentation fault, which is probably not intended. And I believe these features (both clustermesh and auto-direct-node-routes) should be documented much better, from a networking point of view. Like what routes are inserted, what are their types, why can't you read these routes out with a simple routing daemon (like bird or frr), etc. The whole dynamic in-cluster routing is very very shady, and not only requires a ton of knowledge in general networking, but a lot of experimentation, because there is neither a simple guide on how to get this working, nor a page where these large monolithic network options (like auto-direct-node-routes, or ipv4-native-routing-cidr) are explained in more than two sentences (and they do a LOT of things in the background)

**Comment by giorio94** [comment]
> The thing that is strange to me in this question, is when would you actually want auto-direct-node-routes to import remote-nodes as gateways? Only when you have the entire clustermesh on a layer 2 network, which is very rarely the case. 

I think it depends on the use-case, and there may be valid scenarios for both approaches. In case you are interested, I'd suggest you to open a dedicated feature proposal issue to discuss about introducing the support for configuring the direct routes for local nodes only. In general, though, my expectation would be that BGP should be the preferred approach to configure routes in case of more complex configurations, as it provides way more flexibility.

> The "fail silently" was great in our case (even better than excluding remote nodes entirely), because we have some nodes from the other cluster on the same LAN, so those can communicate without packets going through the entire network stack.

I'm not sure I'm following here. Does that imply that the remote cluster doesn't use auto-direct-node-routes at all? Otherwise I would expect the same issue to happen inside the remote cluster as well. That said, the main issue with the fail silent approach is that warnings can be easily missed, and can mislead users in case of problems.

> Also, after this fatal error, cilium throws in a segmentation fault, which is probably not intended.

Could you please share the full error? That doesn't seem to be expected.

**Comment by giorio94** [comment]
> And I believe these features (both clustermesh and auto-direct-node-routes) should be documented much better, from a networking point of view. Like what routes are inserted, what are their types, why can't you read these routes out with a simple routing daemon (like bird or frr), etc. The whole dynamic in-cluster routing is very very shady, and not only requires a ton of knowledge in general networking, but a lot of experimentation, because there is neither a simple guide on how to get this working, nor a page where these large monolithic network options (like auto-direct-node-routes, or ipv4-native-routing-cidr) are explained in more than two sentences (and they do a LOT of things in the background)

Thanks for the feedback. Generally speaking I agree that documentation can always be improved, but OTOH it is quite tricky to document the interaction of multiple features considering all combinations supported by Cilium. That said, tunnel mode is typically considered the simple solution which works out of the box in most setups. Differently, native routing is definitely a bit more complex, and requires some additional customization based on the setup (configuration of routes, native routing CIDR, ...). Would you have time to open a PR to improve some of the parts that you feel would benefit most from an extension?

**Comment by lgyurci** [comment]
> Could you please share the full error? That doesn't seem to be expected.

Sure. There you go:
```
cilium-agent level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: failed to enable direct routes for ipv4: route to destination 10.156.50.37 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.36 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.186.50.25 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.5 contains gateway 10.176.50.254, must be directly reachable\nfailed to enable direct routes for ipv4: route to destination 10.156.50.38 contains gateway 10.176.50.254, must be directly reachable" subsys=daemon
cilium-agent level=info msg="regenerating all endpoints" reason= subsys=endpoint-manager
cilium-agent panic: runtime error: invalid memory address or nil pointer dereference
cilium-agent [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1dee7c1]
cilium-agent
cilium-agent goroutine 160 [running]:
cilium-agent github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Observe(0x3f36470?, {0x3f36588?, 0xc005d82dc0?}, 0x10010008016d756d?, 0x766973756c637845?)
cilium-agent     <autogenerated>:1 +0x21
cilium-agent github.com/cilium/cilium/pkg/stream.First[...]({_, _}, {_, _})
cilium-agent     /go/src/github.com/cilium/cilium/pkg/stream/sinks.go:25 +0x191
cilium-agent github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Get(...)
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/local_node_store.go:145
cilium-agent github.com/cilium/cilium/pkg/node.getLocalNode()
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/address.go:47 +0x9f
cilium-agent github.com/cilium/cilium/pkg/node.GetIPv4()
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/address.go:221 +0x25
cilium-agent github.com/cilium/cilium/pkg/proxy/logger.NewLogRecord({0x3928cce, 0x7}, 0x0?, {0xc001f8dbb0, 0x4, 0xc00667c880?})
cilium-agent     /go/src/github.com/cilium/cilium/pkg/proxy/logger/logger.go:85 +0x158
cilium-agent github.com/cilium/cilium/daemon/cmd.(*Daemon).notifyOnDNSMsg(0xc000c1e380, {0xd90e62?, 0xc002730500?, 0x6094be0?}, 0xc000e34e00, {0xc00287b338, 0x12}, 0x1caa8, {0xc00667c852, 0xe}, ...)
cilium-agent     /go/src/github.com/cilium/cilium/daemon/cmd/fqdn.go:463 +0x1a5c
cilium-agent github.com/cilium/cilium/pkg/fqdn/dnsproxy.(*DNSProxy).ServeDNS(0xc0025145b0, {0x3f4a128, 0xc005a06180}, 0xc0010c5320)
cilium-agent     /go/src/github.com/cilium/cilium/pkg/fqdn/dnsproxy/proxy.go:978 +0x18f6
cilium-agent github.com/cilium/dns.(*Server).serveDNS(0xc000b3ac00, {0xc00093c800, 0x4f, 0x200}, 0xc005a06180)
cilium-agent     /go/src/github.com/cilium/cilium/vendor/github.com/cilium/dns/server.go:653 +0x382
cilium-agent github.com/cilium/dns.(*Server).serveUDPPacket(0xc000b3ac00, 0x0?, {0xc00093c800, 0x4f, 0x200}, {0x3f437e8?, 0xc00144b588}, 0xc000c3a280, {0x0, 0x0})
cilium-agent     /go/src/github.com/cilium/cilium/vendor/github.com/cilium/dns/server.go:605 +0x1cc
cilium-agent github.com/cilium/dns.(*Server).serveUDP.func2()
cilium-agent     /go/src/github.com/cilium/cilium/vendor/github.com/cilium/dns/server.go:531 +0x49
cilium-agent created by github.com/cilium/dns.(*Server).serveUDP in goroutine 952
cilium-agent     /go/src/github.com/cilium/cilium/vendor/github.com/cilium/dns/server.go:530 +0x436
```
> I'm not sure I'm following here. Does that imply that the remote cluster doesn't use auto-direct-node-routes at all?

We have two LANs, and three clusters in the mesh. There is a cluster that has nodes in both of these LANs (lets call it "C", and the other two "A" and "B"). This "C" cluster was using BGP to distribute the routes of its own pods, and as you said, no auto-direct-node-routes. When a pod in cluster "A" tried to reach a pod in cluster "C" directly (due to global services, etc), in some cases, it could do it on layer 2 entirely, when the target pod was on the node running on the same LAN as cluster "A".

> In general, though, my expectation would be that BGP should be the preferred approach to configure routes in case of more complex configurations, as it provides way more flexibility.

Okay, but how? I mean, a full-mesh BGP is not supported by cilium natively, so someone would have to use BIRD or other routing daemons to distribute these routes onto the nodes themselves (like auto-direct-node-routes does), if you'd like to keep the entire cluster communication on layer 2, for performance benefits. Also, using BIRD or FRR to read these routes from the kernel is not possible, due to these routes being "kernel" type, and having a gateway (or at least, that's my guess). So you would have to make a BIRD client run on the nodes (eg. as a daemonset), have cilium estabilish BGP sessions with them, let cilium distribute the pod routes to BIRD, then advertise these pod routes between the BIRDs, and export them into the kernel. And this is a really complicated thing to do, for a very simple thing, that could be easily achievable inside the cilium agent. And if this would be the preferred method, then I would expect some form of documentation for this, hopefully with an example.

Also, we have ended up with this approach approach as a "workaround". Let cilium advertise the routes via BGP into a BIRD instance running on the nodes, distribute the routes between BIRDs via OSPF, and export the routes into the kernel on the nodes. But this just does not feel right, and I would expect the cilium-agent to be able to do this natively.

When I'll have the time, I'll gladly submit feature requests about these.

**Comment by giorio94** [comment]
Thanks for the clarifications. I'll have a look at the panic, that shouldn't happen.

/cc @cilium/sig-bgp for the BGP discussion, I honestly don't have enough context and in-depth knowledge about that part.

**Comment by tommyp1ckles** [comment]
> AFAIK, this scenario is not officially supported at the moment, and `auto-direct-node-routes` applies to both nodes in the local, and possible remote clusters. In other words, there isn't currently the possibility of applying that to a subset of nodes only (although I see that could be useful in certain scenarios). That was already the case in v1.14, although that did only trigger a warning message, and didn't cause a hard failure.
> 
> That said, and independently of this specific use-case, the behavioral change appears to have introduced in [d890e62](https://github.com/cilium/cilium/commit/d890e629f2fd891986166671a47108cf74f7f607) as an effort of bubbling up more errors, although it is not totally clear to me whether the extreme consequence of panic'ing the agent was intended or not. @tommyp1ckles could you please chime in with your thoughts?

The intention of those changes was to improve error handling, specifically without affected any logic (i.e. errors that where only logged before would not be returned early, but appended to a list of errors).

**Comment by tommyp1ckles** [comment]
@giorio94 Looking at where this is failing, it's in the daemon/cmd init stage.  This means that the initial node datapath reconciliation (i.e. `err := d.Datapath().Loader().Reinitialize(...)`).  

Following this, the nodeUpdated call is run from node-manager to reconcile updates to nodes, and those are logged/reported as failures and retried later.

So the issue is that the assumption that we wouldn't fatal on node reconciliations was wrong when performing the first reconciliation.  This was always the [case](https://github.com/cilium/cilium/blob/main/pkg/datapath/loader/base.go#L479-L481), however prior to these changes [this call would never return an error](https://github.com/cilium/cilium/blob/9220ebbb6d1bdca96f6c8327a8f3ba5c37e7bcc7/pkg/datapath/linux/node.go#L1166-L1277).

In the short term, if we want to restore the v1.14 behaviour, we could changes this to log the error instead of returning to fix this regression and prevent other similar ones. However, in the medium term I think a better solution would be to figure how we should handle the remote node route case as described in the thread.

**Comment by tommyp1ckles** [comment]
@giorio94 I've created https://github.com/cilium/cilium/pull/31972 which will return all similar behaviour back to what it was prior to v1.14.  We can backport this to v1.15.

**Comment by giorio94** [comment]
I investigated the panic error you mentioned:

```
cilium-agent panic: runtime error: invalid memory address or nil pointer dereference
cilium-agent [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1dee7c1]
cilium-agent
cilium-agent goroutine 160 [running]:
cilium-agent github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Observe(0x3f36470?, {0x3f36588?, 0xc005d82dc0?}, 0x10010008016d756d?, 0x766973756c637845?)
cilium-agent     <autogenerated>:1 +0x21
cilium-agent github.com/cilium/cilium/pkg/stream.First[...]({_, _}, {_, _})
cilium-agent     /go/src/github.com/cilium/cilium/pkg/stream/sinks.go:25 +0x191
cilium-agent github.com/cilium/cilium/pkg/node.(*LocalNodeStore).Get(...)
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/local_node_store.go:145
cilium-agent github.com/cilium/cilium/pkg/node.getLocalNode()
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/address.go:47 +0x9f
cilium-agent github.com/cilium/cilium/pkg/node.GetIPv4()
cilium-agent     /go/src/github.com/cilium/cilium/pkg/node/address.go:221 +0x25
cilium-agent github.com/cilium/cilium/pkg/proxy/logger.NewLogRecord({0x3928cce, 0x7}, 0x0?, {0xc001f8dbb0, 0x4, 0xc00667c880?})
cilium-agent     /go/src/github.com/cilium/cilium/pkg/proxy/logger/logger.go:85 +0x158
```

That's essentially caused by accessing one of the global getters (`GetIPv4()` in this case) after that the `LocalNodeStore` was stopped, which in turn [resets the global object to nil](https://github.com/cilium/cilium/blob/5d1521bdf3f3f90281a796ee319b651b8a172c08/pkg/node/local_node_store.go#L158). The global getters are meant to be there for backward compatibility only, and eventually replaced as more subsystems get converted to hive. This panic is definitely suboptimal, but OTOH it doesn't cause issues (as the agent is already in the late shutdown phase at that point). Even if a correct error was returned there, a panic would still be triggered by [the `getLocalNode()` method](https://github.com/cilium/cilium/blob/5d1521bdf3f3f90281a796ee319b651b8a172c08/pkg/node/address.go#L48-L53). I personally think it would make sense to put efforts in migrating the remaining subsystems out of the global getters, rather than trying to fix this specific panic occurrence. /cc @joamaki FYI

**Comment by dhedberg** [comment]
Hit by the same issue, but there's no clustermesh involved here.

Our nodes are split over two networks, with the router in between aware of the pod networks on each node (no BGP). Using auto-direct-node-routes with 1.14, cilium would helpfully add direct routes for each node on the same network as one would expect, and while it would obviously fail to add direct routes to the other network that wasn't an issue as those would be handled by the router.

On attempting to upgrade to 1.15, the agents refused to start, as noted in this bug report. We can presumably work around this by just implementing the logic ourselves, but the previous behavior worked better for us.

**Comment by tommyp1ckles** [comment]
Sorry for the delay, looking into this again.  I see that a recent [change](https://github.com/cilium/cilium/pull/32733) set that would've landed in v1.16 has seems to provide a workaround for this.  Specifically the `direct-routing-skip-unreachable` option now modifies that code path to avoid returning errors if that option is set: https://github.com/cilium/cilium/pull/32733/files#diff-14a5efc51ce7ea3959385022d4800b355c766c60e8b7ba0e6052afced05ddbf0R270

**Comment by dhess** [comment]
@tommyp1ckles Thanks! We recently enabled clustermesh on 3 clusters with a `10/8` CIDR and direct routing, and I noticed the "Unable to install direct node route" issue around the same time, but didn't put two and two together until Cilium on one of the clusters finally died. Enabling `direct-routing-skip-unreachable` seems to have fixed the issue.
